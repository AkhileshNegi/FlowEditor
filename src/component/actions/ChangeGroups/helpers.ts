import { Group, ChangeGroups } from '../../../flowTypes';
import { SearchResult } from '../../../store';

export const mapGroupsToSearchResults = (groups: Group[]) =>
    groups.map(({ name, uuid }) => ({ name, id: uuid }));

export const mapSearchResultsToGroups = (searchResults: SearchResult[]) =>
    searchResults.map(result => ({
        uuid: result.id,
        name: result.name
    }));

export const getGroups = (action: ChangeGroups): SearchResult[] => {
    if (action.groups == null) {
        return [];
    }

    if (action.groups.length && action.type !== 'add_contact_groups') {
        return mapGroupsToSearchResults(action.groups);
    }

    return [];
};
