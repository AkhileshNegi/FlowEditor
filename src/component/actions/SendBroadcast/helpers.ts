import { BroadcastMsg, Contact, Group } from '../../../flowTypes';
import { Asset, AssetType } from '../../../services/AssetService';

export const getRecipients = (action: BroadcastMsg): Asset[] => {
    const selected = action.groups.map((group: Group) => {
        return { id: group.uuid, name: group.name, type: AssetType.Group };
    });

    return selected.concat(
        action.contacts.map((contact: Contact) => {
            return { id: contact.uuid, name: contact.name, type: AssetType.Contact };
        })
    );
};
