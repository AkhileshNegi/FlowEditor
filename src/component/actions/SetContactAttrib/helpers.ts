import { Types } from '../../../config/typeConfigs';
import { SetContactField, SetContactProperty } from '../../../flowTypes';
import { snakify, titleCase } from '../../../utils';
import { Asset, AssetType } from '../../../services/AssetService';

export const newFieldAction = (uuid: string, value: string, name: string): SetContactField => ({
    type: Types.set_contact_field,
    field: {
        key: snakify(name),
        name: titleCase(name)
    },
    uuid,
    value
});

export const newPropertyAction = (
    uuid: string,
    value: string,
    name: string
): SetContactProperty => ({
    type: Types.set_contact_property,
    property: snakify(name),
    uuid,
    value
});

export const fieldToAsset = ({ field: { key, name } }: SetContactField): Asset => ({
    id: key,
    name,
    type: AssetType.Field
});

export const propertyToAsset = ({ property }: SetContactProperty): Asset => ({
    id: property,
    name: titleCase(property),
    type: AssetType.Property
});
