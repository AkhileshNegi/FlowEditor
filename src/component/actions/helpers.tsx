import * as React from 'react';

import { Asset, AssetType } from '../../services/AssetService';

const styles = require('../shared.scss');

export const renderAssetList = (assets: Asset[], max: number): JSX.Element[] => {
    return assets.reduce((elements, asset, idx) => {
        if (idx <= max - 1) {
            elements.push(renderAsset(asset));
        } else if (idx > max - 1 && elements.length === max) {
            elements.push(<div key="ellipses">...</div>);
        }
        return elements;
    }, []);
};

export const renderAsset = (asset: Asset) => {
    switch (asset.type) {
        case AssetType.Group:
            return (
                <div className={styles.nodeAsset} key={asset.id}>
                    <span className={`${styles.nodeGroup} icn-group`} />
                    {asset.name}
                </div>
            );
        case AssetType.Label:
            return (
                <div className={styles.nodeAsset} key={asset.id}>
                    <span className={`${styles.nodeLabel} icn-label`} />
                    {asset.name}
                </div>
            );
    }

    return (
        <div className={styles.nodeAsset} key={asset.id}>
            {asset.name}
        </div>
    );
};
