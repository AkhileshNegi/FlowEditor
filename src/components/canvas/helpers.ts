import { FlowPosition } from '~/flowTypes';
import { CanvasPositions } from '~/store/editor';

export const collides = (a: FlowPosition, b: FlowPosition) => {
    // don't bother with collision if we don't have full dimensions
    /* istanbul ignore next */
    if (!a.bottom || !b.bottom) {
        return false;
    }

    return !(b.left > a.right || b.right < a.left || b.top > a.bottom || b.bottom < a.top);
};

export const getCollisions = (
    positions: CanvasPositions,
    box: FlowPosition
): { [uuid: string]: FlowPosition } => {
    const collisions = {};
    for (const nodeUUID of Object.keys(positions)) {
        const position = positions[nodeUUID];
        if (collides(box, position)) {
            collisions[nodeUUID] = position;
        }
    }
    return collisions;
};
