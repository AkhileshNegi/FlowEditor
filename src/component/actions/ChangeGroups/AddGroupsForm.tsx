import { react as bindCallbacks } from 'auto-bind';
import * as isEqual from 'fast-deep-equal';
import * as React from 'react';
import { connect } from 'react-redux';
import { ConfigProviderContext } from '../../../config';
import { ChangeGroups } from '../../../flowTypes';
import { AppState, SearchResult, DispatchWithState } from '../../../store';
import GroupsElement from '../../form/GroupsElement';
import { mapGroupsToSearchResults, mapSearchResultsToGroups } from './helpers';
import ChangeGroupsFormProps from './props';
import { Types } from '../../../config/typeConfigs';
import { bindActionCreators } from 'redux';
import { dump } from '../../../utils';
import AssetService from '../../../services/AssetService';
import { fakePropType } from '../../../config/ConfigProvider';

export interface AddGroupsFormState {
    groups: SearchResult[];
}

export const LABEL = ' Select the group(s) to add the contact to.';
export const PLACEHOLDER = 'Enter the name of an existing group or create a new one';

export const labelSpecId = 'label';

export class AddGroupsForm extends React.PureComponent<ChangeGroupsFormProps, AddGroupsFormState> {
    public static contextTypes = {
        endpoints: fakePropType,
        assetService: fakePropType
    };

    constructor(props: ChangeGroupsFormProps, context: ConfigProviderContext) {
        super(props);

        const groups: SearchResult[] = this.getGroups();

        this.state = {
            groups
        };

        bindCallbacks(this, {
            include: [/^on/]
        });
    }

    public onGroupsChanged(groups: SearchResult[]): void {
        if (!isEqual(groups, this.state.groups)) {
            this.setState({
                groups
            });
        }
    }

    public onValid(): void {
        const newAction: ChangeGroups = {
            uuid: this.props.action.uuid,
            type: this.props.typeConfig.type,
            groups: mapSearchResultsToGroups(this.state.groups)
        };

        this.props.addGroups(this.state.groups);
        this.props.updateAction(newAction);
    }

    private getGroups(): SearchResult[] {
        if (this.props.action.groups === null) {
            return [];
        }

        if (
            this.props.action.groups.length &&
            this.props.action.type !== Types.remove_contact_groups
        ) {
            return mapGroupsToSearchResults(this.props.action.groups);
        }

        return [];
    }

    public render(): JSX.Element {
        return (
            <>
                <p data-spec={labelSpecId}>{LABEL}</p>
                <GroupsElement
                    ref={this.props.onBindWidget}
                    name="Groups"
                    placeholder={PLACEHOLDER}
                    assets={this.context.assetService.getGroupAssets()}
                    groups={this.state.groups}
                    add={true}
                    required={true}
                    onChange={this.onGroupsChanged}
                />
            </>
        );
    }
}

/* istanbul ignore next */
const mapStateToProps = ({ flowContext: { groups }, nodeEditor: { typeConfig } }: AppState) => ({
    groups,
    typeConfig
});

const ConnectedAddGroupForm = connect(mapStateToProps, null, null, { withRef: true })(
    AddGroupsForm
);

export default ConnectedAddGroupForm;
