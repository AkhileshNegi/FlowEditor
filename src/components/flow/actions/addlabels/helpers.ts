import { AddLabelsFormState } from '~/components/flow/actions/addlabels/AddLabelsForm';
import { getActionUUID } from '~/components/flow/actions/helpers';
import { Types } from '~/config/interfaces';
import { AddLabels } from '~/flowTypes';
import { Asset, AssetType } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';

export const initializeForm = (settings: NodeEditorSettings): AddLabelsFormState => {
    if (settings.originalAction && settings.originalAction.type === Types.add_input_labels) {
        const action = settings.originalAction as AddLabels;
        return {
            labels: {
                value: action.labels.map(label => {
                    return { id: label.uuid, name: label.name, type: AssetType.Label };
                })
            },
            valid: true
        };
    }

    return {
        labels: { value: [] },
        valid: false
    };
};

export const stateToAction = (
    settings: NodeEditorSettings,
    formState: AddLabelsFormState
): AddLabels => {
    return {
        type: Types.add_input_labels,
        labels: this.getAsset(formState.labels.value, AssetType.Label),
        uuid: getActionUUID(settings, Types.add_input_labels)
    };
};

export const getAsset = (assets: Asset[], type: AssetType): any[] => {
    return assets.filter((asset: Asset) => asset.type === type).map((asset: Asset) => {
        return { uuid: asset.id, name: asset.name };
    });
};
