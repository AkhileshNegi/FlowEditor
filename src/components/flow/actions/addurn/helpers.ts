import { AddUrnFormState } from '~/components/flow/actions/addurn/AddUrnForm';
import { getActionUUID } from '~/components/flow/actions/helpers';
import { SelectOption } from '~/components/form/select/SelectElement';
import { Scheme, SCHEMES, Types } from '~/config/typeConfigs';
import { AddUrn } from '~/flowTypes';
import { NodeEditorSettings } from '~/store/nodeEditor';

export const getSchemeOptions = (): SelectOption[] => {
    return SCHEMES.map((scheme: Scheme) => {
        return schemeToSelectOption(scheme);
    });
};

export const getSchemeObject = (scheme: string): Scheme => {
    return SCHEMES.find((item: Scheme) => item.scheme === scheme);
};

export const getSchemeSelectOption = (scheme: string): SelectOption =>
    schemeToSelectOption(getSchemeObject(scheme));

export const schemeToSelectOption = (scheme: Scheme): SelectOption => {
    return { value: scheme.scheme, label: scheme.name };
};

export const initializeForm = (settings: NodeEditorSettings): AddUrnFormState => {
    if (settings.originalAction && settings.originalAction.type === Types.add_contact_urn) {
        const { scheme, path } = settings.originalAction as AddUrn;

        return {
            scheme: { value: getSchemeSelectOption(scheme) },
            path: { value: path },
            valid: true
        };
    }

    return {
        scheme: { value: getSchemeSelectOption('tel') },
        path: { value: '' },
        valid: false
    };
};

export const stateToAction = (settings: NodeEditorSettings, formState: AddUrnFormState): AddUrn => {
    return {
        type: Types.add_contact_urn,
        uuid: getActionUUID(settings, Types.add_input_labels),
        scheme: formState.scheme.value.value,
        path: formState.path.value
    };
};
