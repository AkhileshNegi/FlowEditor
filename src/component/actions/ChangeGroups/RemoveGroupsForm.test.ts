import { getTypeConfig } from '../../../config';
import { Types } from '../../../config/typeConfigs';
import { ChangeGroups } from '../../../flowTypes';
import { composeComponentTestUtils, getSpecWrapper, setMock } from '../../../testUtils';
import { createAddGroupsAction } from '../../../testUtils/assetCreators';
import { set, dump } from '../../../utils';
import { labelSpecId } from './AddGroupsForm';
import { mapGroupsToSearchResults, mapSearchResultsToGroups } from './helpers';
import ChangeGroupFormProps from './props';
import {
    LABEL,
    NOT_FOUND,
    PLACEHOLDER,
    REMOVE_FROM_ALL,
    REMOVE_FROM_ALL_DESC,
    RemoveGroupsForm
} from './RemoveGroupsForm';

const addGroupsAction = createAddGroupsAction();
const removeGroupConfig = getTypeConfig(Types.remove_contact_groups);
const removeGroupsAction = {
    ...(addGroupsAction as ChangeGroups),
    type: Types.remove_contact_groups
};

const baseProps: ChangeGroupFormProps = {
    action: removeGroupsAction,
    updateAction: jest.fn(),
    onBindWidget: jest.fn(),
    removeWidget: jest.fn(),
    groups: [],
    typeConfig: removeGroupConfig
};

const { setup, spyOn } = composeComponentTestUtils(RemoveGroupsForm, baseProps);

describe(RemoveGroupsForm.name, () => {
    describe('render', () => {
        it('should render self, children with base props', () => {
            const {
                wrapper,
                instance,
                props,
                // tslint:disable-next-line:no-shadowed-variable
                context: { endpoints }
            } = setup(false, {
                onBindWidget: setMock()
            });
            const label = getSpecWrapper(wrapper, labelSpecId);

            expect(label.is('p')).toBeTruthy();
            expect(label.text()).toBe(LABEL);
            expect(props.onBindWidget).toHaveBeenCalledTimes(2);
            expect(wrapper.find('GroupsElement').props()).toMatchSnapshot();
            expect(wrapper.find('CheckboxElement').props()).toMatchSnapshot();
        });

        it('should render only the checkbox', () => {
            const {
                wrapper,
                instance,
                props,
                // tslint:disable-next-line:no-shadowed-variable
                context: { endpoints }
            } = setup(false, {
                onBindWidget: setMock(),
                removeWidget: setMock()
            });

            instance.onCheck();

            wrapper.update();

            expect(props.removeWidget).toHaveBeenCalledTimes(1);
            expect(props.removeWidget).toHaveBeenCalledWith('Groups');
            expect(getSpecWrapper(wrapper, labelSpecId).exists()).toBeFalsy();
            expect(wrapper.find('GroupsElement').exists()).toBeFalsy();
            expect(wrapper.find('CheckboxElement').props()).toEqual({
                name: REMOVE_FROM_ALL,
                defaultValue: true,
                description: REMOVE_FROM_ALL_DESC,
                sibling: false,
                onCheck: instance.onCheck
            });
        });
    });

    describe('instance methods', () => {
        describe('getGroups', () => {
            it("should return an empty list if action's groups are null", () => {
                const grouplessAction = { ...removeGroupsAction, groups: null };
                const { wrapper, instance } = setup(true, {
                    action: set(grouplessAction)
                });
                const returnedGroups = instance.getGroups();

                expect(returnedGroups).toEqual([]);
                expect(returnedGroups).toMatchSnapshot();
            });

            it("should return an empty list if action's groups property is an empty list", () => {
                const grouplessAction = { ...removeGroupsAction, groups: [] };
                const { wrapper, instance } = setup(true, {
                    action: set(grouplessAction)
                });
                const returnedGroups = instance.getGroups();

                expect(returnedGroups).toEqual([]);
                expect(returnedGroups).toMatchSnapshot();
            });

            it('should return empty list if action is add groups action', () => {
                const { wrapper, instance, props: { action } } = setup(true, {
                    action: set(addGroupsAction)
                });
                const returnedGroups = instance.getGroups();

                expect(returnedGroups).toEqual([]);
                expect(returnedGroups).toMatchSnapshot();
            });

            it('should return SearchResult[] if action is remove groups action and it has groups', () => {
                const { wrapper, instance, props } = setup();
                const searchResults = mapGroupsToSearchResults(props.action.groups);
                const returnedGroups = instance.getGroups();

                expect(returnedGroups).toEqual(searchResults);
                expect(returnedGroups).toMatchSnapshot();
            });
        });

        describe('onGroupsChanged', () => {
            it('should update groups state if passed new groups', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, instance, props } = setup();
                const searchResults = mapGroupsToSearchResults(props.action.groups.slice(2));

                instance.onGroupsChanged(searchResults);

                expect(setStateSpy).toHaveBeenCalledTimes(1);
                expect(setStateSpy).toHaveBeenCalledWith({ groups: searchResults });

                setStateSpy.mockRestore();
            });

            it('should not update groups state if passed same groups', () => {
                const setStateSpy = spyOn('setState');
                const { wrapper, instance, props } = setup();
                const searchResults = mapGroupsToSearchResults(props.action.groups);

                expect(wrapper.state('groups')).toEqual(searchResults);

                instance.onGroupsChanged(searchResults);

                expect(setStateSpy).not.toHaveBeenCalled();

                setStateSpy.mockRestore();
            });
        });

        describe('onValid', () => {
            it('should call updateAction action creator with a ChangeGroups action', () => {
                const {
                    wrapper,
                    instance,
                    props: { action, updateAction: updateActionMock }
                } = setup(true, { updateAction: setMock() });
                const expectedAction = {
                    uuid: action.uuid,
                    type: action.type,
                    groups: mapSearchResultsToGroups(wrapper.state('groups'))
                };
                instance.onValid();
                expect(updateActionMock).toHaveBeenCalledWith(expectedAction);
            });

            it('should honor the remove all flag', () => {
                const {
                    wrapper,
                    instance,
                    props: { action, updateAction: updateActionMock }
                } = setup(true, { updateAction: setMock() });

                // set the remove all flag
                wrapper.setState({ removeFromAll: true });

                // our action shouldn't include any groups
                const expectedAction = {
                    uuid: action.uuid,
                    type: action.type,
                    groups: []
                };

                instance.onValid();
                expect(updateActionMock).toHaveBeenCalledWith(expectedAction);
            });
        });
    });
});
