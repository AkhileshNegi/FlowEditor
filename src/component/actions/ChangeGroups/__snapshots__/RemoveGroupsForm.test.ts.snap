// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveGroupsForm instance methods getGroups should return SearchResult[] if action is remove groups action and it has groups 1`] = `
Array [
  Object {
    "id": "23ff7152-b588-43e4-90de-fda77aeaf7c0",
    "name": "Customers",
    "type": "group",
  },
  Object {
    "id": "2429d573-80d7-47f8-879f-f2ba442a1bfd",
    "name": "Unsatisfied Customers",
    "type": "group",
  },
  Object {
    "id": "cdbf9e01-aaa7-4381-8259-ee042447bcac",
    "name": "Early Adopters",
    "type": "group",
  },
  Object {
    "id": "afaba971-8943-4dd8-860b-3561ed4f1fe1",
    "name": "Testers",
    "type": "group",
  },
  Object {
    "id": "33b28bac-b588-43e4-90de-fda77aeaf7c0",
    "name": "Subscribers",
    "type": "group",
  },
]
`;

exports[`RemoveGroupsForm instance methods getGroups should return an empty list if action's groups are null 1`] = `Array []`;

exports[`RemoveGroupsForm instance methods getGroups should return an empty list if action's groups property is an empty list 1`] = `Array []`;

exports[`RemoveGroupsForm instance methods getGroups should return empty list if action is add groups action 1`] = `Array []`;

exports[`RemoveGroupsForm render should render self, children with base props 1`] = `
Object {
  "add": false,
  "assets": GroupAssets {
    "assetType": "group",
    "assets": Object {},
    "endpoint": "/assets/groups.json",
    "idProperty": "uuid",
    "localStorage": true,
    "nameProperty": "name",
  },
  "groups": Array [
    Object {
      "id": "23ff7152-b588-43e4-90de-fda77aeaf7c0",
      "name": "Customers",
      "type": "group",
    },
    Object {
      "id": "2429d573-80d7-47f8-879f-f2ba442a1bfd",
      "name": "Unsatisfied Customers",
      "type": "group",
    },
    Object {
      "id": "cdbf9e01-aaa7-4381-8259-ee042447bcac",
      "name": "Early Adopters",
      "type": "group",
    },
    Object {
      "id": "afaba971-8943-4dd8-860b-3561ed4f1fe1",
      "name": "Testers",
      "type": "group",
    },
    Object {
      "id": "33b28bac-b588-43e4-90de-fda77aeaf7c0",
      "name": "Subscribers",
      "type": "group",
    },
  ],
  "name": "Groups",
  "onChange": [Function],
  "placeholder": "Enter the name an existing group",
  "required": true,
  "searchPromptText": "Enter the name of an existing group",
}
`;

exports[`RemoveGroupsForm render should render self, children with base props 2`] = `
Object {
  "defaultValue": false,
  "description": "Remove the active contact from all groups they're a member of.",
  "name": "Remove from All",
  "onCheck": [Function],
  "sibling": true,
}
`;
